Lms/normalize concept results

anathomical opened the pull request:
## WHAT
Add a suite of models to support the normalization of `ConceptResult` records along with code to actually handle that migration
## WHY
We want to slow the rapid increase in database storage necessitated by the current denormalized and highly-repeated nature of current `ConceptResult` records.  This new approach should vastly reduce the amount of required space for the same data.
## HOW
- Create a series of migrations to create all the necessary tables we'll be using for denormalization
- Define models with basic tests and validation for all of the models
- Ensure that the root model (`StudentResponse`) can create new normalized records from either existing `ConceptResult` records, or from the JSON payload that's currently used by `ActivitySession`s to create new `ConceptResult`s.
- Ensure that the root model can present itself in the same shape as the old `ConceptResult` records so that we can ultimately use the new model as a drop-in replacement
- Create a series of Sidekiq jobs to copy existing `ConceptResult` records into the new `StudentResponse` model

### Notion Card Links
https://www.notion.so/quill/ConceptResult-RFC-Implementation-33dd662d331a4a0896b1a5bf00f3298d
https://www.notion.so/quill/ConceptResults-Pure-PostgreSQL-Normalization-RFC-b0fd4252f72c4ef68e10d91e5a52afca
https://www.notion.so/quill/Determine-ConceptResult-metadata-schema-3d1b92ff70b8435db5752c00acf78eb6

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Deploying now
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

anathomical committed:
Begin skeletonizing normalized concept result models

anathomical committed:
Some more updates on handling extra metadata keys

anathomical committed:
Commit annotation changes on older models from migration

anathomical committed:
Add migrations for all new models used in for ConceptResult normalization

anathomical committed:
Update models with additional code and refactor to use concern

anathomical committed:
Add factories and initial validation tests for new models

anathomical committed:
Fix linting errors

anathomical committed:
Fix broken tests

anathomical committed:
Rename function to something more reasonable

anathomical committed:
Make some classmethods private, scaffold method tests

anathomical committed:
Add additional models and migrations

anathomical committed:
Add PreviousFeedbackText to StudentResponse

anathomical committed:
Use jsonb for AnswerText since sometimes it's an array or hash

anathomical committed:
Add question_score to normalized model, update tests

anathomical committed:
Get all tests written and working

anathomical committed:
Do a bunch of refactoring for linting and simplification

anathomical committed:
Some clean-up around using "text" for answer normalization

anathomical committed:
Add Sidekiq workers for migrating ConceptResults to StudentResponses

anathomical committed:
Add a small comment

anathomical committed:
Remove unused model file

anathomical committed:
Fix linting errors

happythenewsad commented on services/QuillLMS/app/models/concept_result.rb:
```diff
@@ -29,6 +29,8 @@ class ConceptResult < ApplicationRecord
   validates :concept, presence: true
   validates :activity_session_id, presence: true
 
+  has_one :student_response_concept_result, dependent: :destroy
```
This is to avoid adding a foreign key column to a 1.4B row table, right? 👍

anathomical replied:
I hadn't fully thought that through, actually, but that is a very good reason not to put a ForeignKey constraint in there.

happythenewsad commented on services/QuillLMS/app/models/student_response.rb:
```diff
@@ -0,0 +1,255 @@
+  # that this model knows how to normalize.  Any keys that are
+  # fed into create_from_json that aren't from this list,
+  # will be stashed in a related StudentResponseExtraMetadata
+  # record
```
If we're relegating the extra metadata to a separate table, why not throw it away entirely?

anathomical replied:
This is a fair question.  I think a good chunk of this is a sense of paranoia, but on the more practical side of things, one of the things we're planning to do is test this extensively in staging.  Doing so should allow us to use the `extra_metadadata` table to ACTUALLY inventory the keys that we don't currently support.  The hope is that there will be a lot fewer rows in that table than in the full `concept_results` table because most of the `ConceptResults` records should transfer into `StudentResponse` records that have no `extra_metadata`.  We may, once we take a look at what sorts of data ends up in `extra_metadata`, drop the table.  We just want to make sure we look at it first.

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/normalize-concept-results

happythenewsad commented on services/QuillLMS/spec/models/student_response_spec.rb:
```diff
@@ -0,0 +1,345 @@
+
+      it 'should not link to records when the appropriate keys are not provided' do
+        response = StudentResponse.create_from_json(json)
+        metadata = json[:metadata]
```
remove unnecessary line

anathomical replied:
Good catch.  Done.

happythenewsad commented on services/QuillLMS/spec/workers/enqueue_concept_results_migration_worker_spec.rb:
```diff
@@ -0,0 +1,54 @@
+      subject.perform(nil, nil)
+    end
+
+    it 'should take start params to skip existing ConceptResults' do
```
Nit: These two specs would provide more test coverage if they actually filtered out ConceptResults via the values of `start` and `finish`.

anathomical replied:
So, secretly they both are because the `context` block includes a `let!` that creates a `concept_result`.  That means that there are two in the DB, and the test conditions check to ensure only one is getting processed, but that's definitely not obvious when I look at them again.  I think it's reasonable to tweak the tests to make it more obvious that they are testing the conditions that they are to avoid confusion.

happythenewsad reviewed (changes requested):
Solid!

anathomical committed:
Clean up some dirty db/structure stuff from a bad local migration

anathomical committed:
Make it more obvious what exactly we're testing with start/finish params

anathomical committed:
Remove extraneous line

anathomical committed:
Allow student_response.question_id to be null

Some of our historical ConceptResult records do not have the information we need to link the ConceptResult to a question

dandrabik commented on services/QuillLMS/app/models/student_response.rb:
```diff
@@ -0,0 +1,255 @@
+    StudentResponseExtraMetadata.new(metadata: extra_metadata)
+  end
+
+  private_class_method def self.set_normalized_text(target, data_hash)
```
shouldn't this be an instance method?

anathomical replied:
Seems reasonable.  Will change.

dandrabik commented on services/QuillLMS/app/models/student_response.rb:
```diff
@@ -0,0 +1,255 @@
+class StudentResponse < ApplicationRecord
+  belongs_to :activity_session
+  belongs_to :question
+  belongs_to :student_response_answer_text
```
Do we need `_text` in all of these names? Could this be `student_response_answer`? Also, I'm not sure `student` is needed in these names as well. Maybe you are trying to avoid a similar name to the CMS, but it might be better to  call all of these `response`, `response_answer`, etc. It would make the code a lot more brief and increase readability. We should discuss.

anathomical replied:
This seems like a totally reasonable suggestion.  I was uncomfortable with the long names, but already had it in my head, so couldn't come up with a better alternative.  I'm going to implement this suggestion since I think you're right that it's more readable.

dandrabik commented on services/QuillLMS/app/models/student_response.rb:
```diff
@@ -0,0 +1,255 @@
+      student_response_extra_metadata: extra_metadata
+    )
+
+    set_normalized_text(student_response, data_hash)
```
Seems like `set_normalized_text` should be an instance method.

anathomical replied:
Seems reasonable.  Will make that change.

dandrabik commented on services/QuillLMS/app/models/concerns/is_student_response_normalized_text.rb:
```diff
@@ -0,0 +1,15 @@
+  included do
+    has_many :student_responses
+
+    validates :text, uniqueness: true, length: { minimum: 0, allow_nil: false }
```
Why does this validate a minimum length of zero?

anathomical replied:
I am (perhaps incorrectly) given to understand that this is how you validate non-null values while allowing an empty string.

anathomical committed:
Tweak StudentResponse to handle missing keys in old ConceptResults

dandrabik commented on services/QuillLMS/app/models/concerns/is_student_response_normalized_text.rb:
```diff
@@ -0,0 +1,15 @@
+# This concern is intended to describe the many different tables that we use
+# to normalize different types of text.  Basically these tables all do the
+# same thing, but we wanted to keep the various concerns separated.
+module IsStudentResponseNormalizedText
```
I'm on the fence about this module. I'm not sure if this abstraction is necessary, given how little code is here. It might be clearer to follow for future developers to just have the two lines in each model.

anathomical replied:
I was on the fence about it myself.  I'm totally glad to just put the two lines in the model files.  Especially with the comments about `json` fields you left elsewhere.  If we're not going to use the abstract in all six models, we might as well just be explicit in them instead.

dandrabik commented on services/QuillLMS/app/models/student_response.rb:
```diff
@@ -0,0 +1,257 @@
+  end
+
+  private_class_method def self.set_normalized_text(target, data_hash)
+    metadata = data_hash[:metadata]
```
Would it be better to just pass in `metadata` to this method?

anathomical replied:
Not entirely obvious given the repetition, but while we pull 5 of our six values out of `metadata`, the value for `question_type` is expected to come from `data_hash` directly.

dandrabik commented on services/QuillLMS/app/models/student_response.rb:
```diff
@@ -0,0 +1,257 @@
+    # six strings we care about without actually writing the logic six
+    # different times.
+
+    attribute_model_value_tuples = [
```
So I wonder if the long names are driving this meta-programming. If 1) this were an instance method (as it should be), 2) the names were shorter, and 3) the texts had validations for empty values, the non-meta-programming equivalent of this code could be:
```ruby
self.response_answer = ResponseAnswer.find_or_create(text: metadata[:answer])
self.response_direction = ResponseDirection.find_or_create(text: metadata[:directions])
self.response_instruction = ResponseInstruction.find_or_create(text: metadata[:instructions])
...
```
There's duplication, but that would be a pretty straightforward method to follow.

anathomical replied:
I did want to guard against `nil` values, but I can just do that with guard statements.  I'll update to your suggested approach.

dandrabik commented on services/QuillLMS/app/models/student_response.rb:
```diff
@@ -0,0 +1,257 @@
+  end
+
+  private def legacy_format_base_metadata
+    metadata = {
```
This `metadata=` is not used.
```suggestion
     {
```

anathomical replied:
Good catch.  Removed.

dandrabik commented on services/QuillLMS/app/models/student_response.rb:
```diff
@@ -0,0 +1,257 @@
+  # and rolls them up so that a single response has only one entry
+  # plus an array of concept_ids that match to it
+  private_class_method def self.roll_up_concepts(data_hash_array)
+    data_hash_array.each_with_object({}) do |value, rollup|
```
I'm slightly confused here by the specific `key` construction that you are later throwing away with `.values`, but if this is your approach, I think you want to look at `.group_by`.

This is equivalent
```ruby
data_hash_array
  .group_by {|value| [value[:metadata][:questionNumber], value[:metadata][:attemptNumber]]}
  .map {|_,value| {concept_ids: value.map {|v| v[:concept_id}}
```
And returns something of the shape `[{concept_ids: [1,2,3]}, {concept_ids: [1,2,3]}]`. However, this shape isn't what the comment above it states. If it doesn't have them let's add a test for this method.

anathomical replied:
I think you might be missing that on line 219 we pull `value` into the rollup value.  This behavior is covered in a test (`student_response_spec.rb:157`).

anathomical replied:
That said, I think that `group_by` is a much cleaner approach that the stuff I'm doing with `key`, so I think I can convert the code I have here to something using that approach.  Good suggestion.  (I wasn't aware of `group_by`, but it's definitely much cleaner for this application.)

dandrabik commented on services/QuillLMS/app/models/student_response.rb:
```diff
@@ -0,0 +1,257 @@
+  end
+
+  private def legacy_format_metadata
+    metadata = legacy_format_base_metadata
```
To do this in an immutable way, consider the equivalent code:
```ruby
legacy_format_base_metadata
  .merge(student_response_extra_metadata || {})
  .reject { |_,v| v.blank? }

anathomical replied:
Seems reasonable.  Will update.

dandrabik commented on services/QuillLMS/db/migrate/20220505154357_create_student_response_answer_text.rb:
```diff
@@ -0,0 +1,18 @@
+
+class CreateStudentResponseAnswerText < ActiveRecord::Migration[5.1]
+  def change
+    create_table :student_response_answer_texts do |t|
```
I think this is really surprising (and I think you agree based on this 6-line comment). We should not do this. I think trying to abstract this has led us a bit astray here.

anathomical replied:
Went back to calling this column `json`

anathomical committed:
Allow StudentResults to be associated with multiple ConceptResults

Since we're normalizing multiple concepts into a single record

dandrabik commented on services/QuillLMS/app/workers/copy_concept_results_to_student_responses_worker.rb:
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class CopyConceptResultsToStudentResponsesWorker
```
What's the case for the job looping through a batch of `concept_ids` vs. 1 job per `concept_id`? This isn't idempotent, so wouldn't we end up with duplicate `student_responses` if this job errors midway through?

anathomical replied:
It may be reasonable to rework this from a batch executor to an individual executor just to keep things simple.  It sort of felt to me like putting fewer jobs on the queue might be a benefit, but on second thought I'm not totally sure that's true.

As for idempotency, `StudentResponse.create_from_concept_result` itself should be idempotent (there are tests around that), so if some items in a batch fail, rerunning the whole batch should be safe.

dandrabik commented on services/QuillLMS/db/migrate/20220505155016_create_student_response.rb:
```diff
@@ -0,0 +1,22 @@
+      t.references :question
+      t.integer :question_number
+      t.float :question_score
+      t.references :student_response_answer_text, foreign_key: true
```
Do we have any concerns that these foreign keys will cause performance issues on this giant table. FK's do slow down inserts and they often removed from very, very large tables for performance reasons. Probably worth doing a little bit of research on this and/or keeping an eye on this if you do keep the foreign keys.

dandrabik commented on services/QuillLMS/spec/factories/student_response_answer_text.rb:
```diff
@@ -0,0 +1,7 @@
+
+FactoryBot.define do
+  factory :student_response_answer_text, class: 'StudentResponseAnswerText' do
+    sequence(:text) { |n| "This a student response answer #{n}." }
```
I think the naming of this field even tricked yourself (you have this as a `jsonb` column). Again, let's change that name.

anathomical replied:
Name updated.

dandrabik commented on services/QuillLMS/app/models/student_response.rb:
```diff
@@ -0,0 +1,257 @@
+      question_number: concept_result.metadata['questionNumber']
+    )
+
+    StudentResponse.transaction do
```
I'd be curious with this transaction and the logic, how long with it take to run 1.7 billion of these?

dandrabik commented on services/QuillLMS/app/models/student_response.rb:
```diff
@@ -0,0 +1,257 @@
+    # conept_ids if that hasn't been done by an earlier step
+    data_hash[:concept_ids] = [data_hash[:concept_id]] unless data_hash.key?(:concept_ids) && !data_hash[:concept_ids].empty?
+
+    extra_metadata = extract_extra_metadata(metadata)
```
There's a mix of class methods/variables, that likely should be instance variables/methods, e.g. you might consider setting an `attr_accessor` of `metadata` on the student_response model, then you could set everyone on the instance from that.

dandrabik reviewed (changes requested):
Thanks for all of your work on this. I left a good amount of comments, so it might be worth talking through them together.

Some of the major items:
1. Renaming the tables to be a little more succinct.
2. Remove some abstraction from a small mixin
3. Remove some metaprogramming in favor of slightly duplicative, but simple code.

Plus a bunch of smaller comments. It's worth reading through first, then we should chat.

anathomical committed:
Update 'should' text on test to be accurate from copy/paste

anathomical committed:
Rename all of our models to be shorter and simpler

anathomical committed:
Tweak some of our logic to be less meta-program-y and more Rails-y

anathomical committed:
Remove normalized text concern abstraction

anathomical committed:
Do some minor refactors based on Dan's suggestions in PR

anathomical committed:
Make a couple of small tweaks based on feedback

anathomical committed:
Refactor migration workers to be more efficient

anathomical committed:
Fix linting errors

anathomical committed:
Update db documentation and db/structure file

anathomical committed:
Fix linting errors, stop saving empty-string normalized text

anathomical committed:
Manually remove command that's invalid in older PostgreSQL versions

anathomical committed:
Update annotation and db/structure with pg_dump v10

anathomical committed:
Remove foreign_key constraints from Response

Add unique index to 'json' column for ResponseAnswer

These are both done in the hopes that it'll speed up migration rates

anathomical committed:
Add benchmarking for Response migrations

Also update tests for normalized text to confirm validation of minimum text length of 1

anathomical committed:
Update batch migration worker to do less up-front work

anathomical committed:
Try removing transaction block to test speed

anathomical committed:
Add transactions back

anathomical committed:
Some minor tweaks and clean-up

anathomical committed:
Fix linting errors

anathomical committed:
Quick and dirty changes to speed up migration testing

anathomical committed:
Remove extraneous indexes

anathomical committed:
Disable validation for cached normalized strings when migrating

anathomical committed:
Refactor to use insert-fetch-id to reduce round trips in migration

anathomical committed:
Tweak insert-get-id query slightly

anathomical committed:
Refactor to use non-normalized Answer to see if it migrates faster

anathomical committed:
Update db annotations

anathomical committed:
Remember that answers are json, not always text

anathomical committed:
Fix minor bug that failed to set direction normalization references

anathomical committed:
Remove a bunch of models that no longer exist

anathomical committed:
Update tests to match new behavior

anathomical committed:
Update migrations to use fewer BIGINTs for IDs

anathomical committed:
Merge branch 'lms/normalize-concept-results-non-normalized-answers' into lms/normalize-concept-results

anathomical committed:
Update workers to batch as intended

anathomical committed:
Move Repsonse migration to its own sidekiq queue

anathomical committed:
Tweak enqueue job to not overlap finish and subsequent start

anathomical committed:
Turn of CyclomaticComplexity linting for migration

This code shouldn't need maintenance moving forward, and keeping it all in one block is, I think, a benefit in this case

happythenewsad commented on services/QuillLMS/app/models/response.rb:
```diff
@@ -0,0 +1,145 @@
+  # will be stashed in a related ResponseExtraMetadata
+  # record
+  KNOWN_METADATA_KEYS = [
+    :answer,
```
Were any of these keys always present, during your sampling? if so, we can make our data shape stricter by enforcing `null: false` in the migration file. Even if some are not always present, or if we're not 100% confident since we didn't sample the entire ConceptResult table, we could add sane defaults.

anathomical replied:
All of the text normalization handles nulling sort of automatically via foreign keys, but do you have thoughts on `NOT NULL`ing any of the other columns?

happythenewsad commented on services/QuillLMS/app/models/response_instructions.rb:
```diff
@@ -0,0 +1,19 @@
+# Table name: response_instructions
+#
+#  id         :integer          not null, primary key
+#  text       :text             not null
```
Does a not null text column default to `''`?

anathomical replied:
Not unless specified.  Basically there must be a specified string here, never an inferred one.  If a `Response` doesn't have (in this case) an `instructions` value, no `ResponseInstructions` will be created, and the foreign key reference in the `Response` will be `NULL`, not pointing to a non-existent value.

happythenewsad commented on services/QuillLMS/app/workers/enqueue_concept_results_migration_worker.rb:
```diff
@@ -0,0 +1,19 @@
+
+  BATCH_SIZE=100000
+
+  def perform(start, finish)
```
This could have been a rake task. Personal preference, I guess.

happythenewsad commented on services/QuillLMS/app/workers/copy_concept_results_to_responses_worker.rb:
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+class CopyConceptResultsToResponsesWorker
```
Please add a note in your rollout strategy to actively monitor the Sidekiq UI for job failures. We'll need to notice these and also capture the args for the failed job, in order to know what to re-run.

happythenewsad reviewed (approved):
Approving, assuming that my comments about null and default column values have already been anticipated.

brendanshean commented on services/QuillLMS/db/migrate/20220505154951_create_response_prompt.rb:
```diff
@@ -0,0 +1,15 @@
+class CreateResponsePrompt < ActiveRecord::Migration[5.1]
+  def change
+    create_table :response_prompts, id: false do |t|
+      t.serial :id, limit: 4, primary_key: true
```
What's the reason for using `t.serial` here?

brendanshean replied:
Nevermind. I use `create_table :response_prompts, id: :integer` for `limit: 4` and then `:bigint` for `limit: 8`. I had not seen `t.serial` before.

anathomical replied:
I hadn't either until I was trying to figure out how to specify auto incremented IDs directly.  The internet taught me.

brendanshean commented on services/QuillLMS/app/workers/copy_concept_results_to_responses_worker.rb:
```diff
@@ -0,0 +1,54 @@
+        prompt = concept_result.metadata['prompt']
+        question_type = concept_result.question_type
+
+        directions_cache[directions] ||= ResponseDirections.find_or_create_by(text: directions)&.id if directions
```
Are we concerned with any `find_or_create_by` silently failing?  I realize the data is already in the database, but the unknown unknowns always lurk, especially with so many records..  Maybe `add a find_or create_by!`  and a corresponding `rescue` clause that documents the failed creation?

anathomical replied:
What would a failure look like?  Are you imagining that it somehow fails to find, tries to do a create, but then the `create` fails?  That seems like a reasonable thing to worry about.  I think we could just get away with the just making it `find_or_create_by!` and not rescuing.  As long as we're monitoring the retry queue, we should be able to keep things cleaned up.

brendanshean replied:
That sounds good.  I just realized that `rescue` clauses add a performance penalty, and given the size of this migration, it's probably better to just monitor the retry queue like you suggest.

brendanshean commented on services/QuillLMS/spec/workers/copy_concept_results_to_responses_worker_spec.rb:
```diff
@@ -0,0 +1,80 @@
+        subject.perform(concept_result.id, concept_result.id)
+        subject.perform(concept_result.id, concept_result.id)
+      end.to change(Response, :count).by(1)
+         .and change(Response, :count).by(1)
```
What's the purpose of this conjunct?

anathomical replied:
Good catch.  I think that may have been a different value in some other version of the approach, but it is definitely unnecessary now.

brendanshean reviewed (approved):
Great work on this!   Nice clear testing and separation of new concepts.  I left a few comments but nothing major that should hold up approval.

anathomical committed:
Make some tweaks based on PR feedback

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/normalize-concept-results

dandrabik commented on services/QuillLMS/app/models/response.rb:
```diff
@@ -0,0 +1,145 @@
+
+  private def legacy_format_metadata
+    legacy_format_base_metadata
+      .merge!(extra_metadata || {})
```
This should use the non-destructive merge to maintain immutability
```suggestion
      .merge(extra_metadata || {})
```

dandrabik commented on services/QuillLMS/app/models/response.rb:
```diff
@@ -0,0 +1,145 @@
+  belongs_to :response_prompt
+  belongs_to :response_question_type
+
+  validates_exclusion_of :correct, in: [nil]
```
Since this is a boolean, would it be more explicit to validate this way?:
```ruby
validates :correct, inclusion: {in: [true, false]}

dandrabik commented on services/QuillLMS/app/models/response_directions.rb:
```diff
@@ -0,0 +1,19 @@
+class ResponseDirections < ApplicationRecord
+  has_many :responses
+
+  validates :text, uniqueness: true, length: { minimum: 1, allow_nil: false }
```
I believe your length validations are equivalent to the standard `presence: true`, e.g.
```suggestion
  validates :text, uniqueness: true, presence: true
```

dandrabik commented on services/QuillLMS/app/models/response_instructions.rb:
```diff
@@ -0,0 +1,19 @@
+class ResponseInstructions < ApplicationRecord
+  has_many :responses
+
+  validates :text, uniqueness: true, length: { minimum: 1, allow_nil: false }
```
Note the `presence: true` applies to all of these

dandrabik commented on services/QuillLMS/app/workers/copy_concept_results_to_responses_worker.rb:
```diff
@@ -0,0 +1,54 @@
+
+        extra_metadata = Response.parse_extra_metadata(concept_result.metadata)
+
+        worker.add({
```
Let's name this something other than `worker`, since it's slightly confusing within a file named `_worker`. Maybe `bulk_insert`?

dandrabik commented on services/QuillLMS/app/workers/copy_concept_results_to_responses_worker.rb:
```diff
@@ -0,0 +1,54 @@
+    question_types_cache = {}
+
+    Response.bulk_insert(ignore: true) do |worker|
+      ConceptResult.where(id: start..finish).each do |concept_result|
```
This will load 100,000 records into memory, but only working on one at a time. We should use [find_each](https://apidock.com/rails/ActiveRecord/Batches/find_each) to find this in smaller batches (even something like 10,000 records). Or you could adjust the batch size.

dandrabik commented on services/QuillLMS/app/workers/copy_concept_results_to_responses_worker.rb:
```diff
@@ -0,0 +1,54 @@
+  include Sidekiq::Worker
+  sidekiq_options queue: SidekiqQueue::MIGRATION
+
+  BATCH_SIZE=100000
```
Small style nit, you can use the `_` symbol in Ruby integers to make large ones more readable, e.g.
```suggestion
  BATCH_SIZE = 100_000
```

dandrabik commented on services/QuillLMS/app/models/response.rb:
```diff
@@ -0,0 +1,145 @@
+#
+#  index_responses_on_concept_result_id  (concept_result_id) UNIQUE
+#
+class Response < ApplicationRecord
```
Naming-wise, does response still make sense given how the scope has changed? Should this be something like `concept_results`?

dandrabik reviewed (changes requested):
Overall this looks good. A bunch of small comments for tweaks to make.

The one item I'd like to discuss is the naming of these tables. I believe with this new shape we discussed naming them something along the lines of `concept_results` (and naming the old table `concept_results_legacy` or something). Given it affects so much code, we should lock that down before rolling this out. Let's chat tomorrow morning.

anathomical committed:
Some minor clean-up based on PR feedback

anathomical committed:
Modify migration worker to work in batches

So that we avoid loading too many records from SQL into memory

brendanshean commented on services/QuillLMS/spec/rails_helper.rb:
```diff
@@ -34,6 +34,10 @@
+# Define any negatated matchers we need
+# https://relishapp.com/rspec/rspec-expectations/v/3-8/docs/define-negated-matcher
+RSpec::Matchers.define_negated_matcher :not_change, :change
+
```
@anathomical I'm really happy you found this.  I've started using it elsewhere.

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/normalize-concept-results

anathomical committed:
Update migrations for new table names

anathomical committed:
Do a bunch of work to rename `Response` to `ConceptResult`

anathomical committed:
Update migration workers to reference new names for models and relations

brendanshean commented on services/QuillLMS/app/models/concept_result.rb:
```diff
@@ -100,19 +100,19 @@ def legacy_format
-    self.response_previous_feedback = ResponsePreviousFeedback.find_or_create_by(text: metadata[:lastFeedback])
-    self.response_prompt = ResponsePrompt.find_or_create_by(text: metadata[:prompt])
-    self.response_question_type = ResponseQuestionType.find_or_create_by(text: data_hash[:question_type])
+    self.concept_result_directions = ConceptResultDirections.find_or_create_by(text: metadata[:directions])
```
Probably too late with the upcoming Rails 6 upgrade, but something to be aware of in the future; 
 https://www.bigbinary.com/blog/rails-6-adds-create_or_find_by

brendanshean commented on services/QuillLMS/spec/factories/concept_result.rb:
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :concept_result, class: 'ConceptResult' do
```
In all of the subsequent factory definitions, I don't think it's necessary to explicitly name `class: [ClassName]`.  Factory bot should infer it [automatically](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#specifying-the-class-explicitly).

anathomical replied:
This probably is true, but there are, like, 15 or so examples in our Factory files that use this pattern.  I was trying to maintain that pattern.  However, another 60 or so factories don't follow that pattern.  Maybe it's worth creating a card to just generally clean this up to unify the approach?

dandrabik replied:
If the factory is the same name snakecase as the class, then `class:` is extraneous. Let's remove the `class:` flags from this PR. 

Note, there are plenty of non-optimal patterns in the repo, but PRs are the proper place to ensure we are following proper patterns for new code.

brendanshean replied:
@anathomical I have a side [branch](https://github.com/empirical-org/Empirical-Core/tree/bs-standardize-factory-names) for factory cleanup so I can remove those other ones there.

anathomical committed:
Ensure that we gracefully handle empty string normalization

dandrabik commented on services/QuillLMS/app/models/concept_result.rb:
```diff
@@ -0,0 +1,145 @@
+  end
+
+  def self.parse_extra_metadata(metadata)
+    metadata.deep_symbolize_keys.except(*KNOWN_METADATA_KEYS).reject{ |_,v| v.nil? || (v.is_a?(Enumerable) && v.empty?) }
```
Let's put code like this on multiple lines for readability
```suggestion
    metadata
      .deep_symbolize_keys
      .except(*KNOWN_METADATA_KEYS)
      .reject{ |_,v| v.nil? || (v.is_a?(Enumerable) && v.empty?) }
```

dandrabik reviewed (approved):
A few small suggestions, but overall :+1:. Thanks for putting this together.

anathomical committed:
Clean up factory definitions so that extraneous stuff isn't there

anathomical committed:
Do some clean-up based on PR feedback

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/normalize-concept-results

anathomical committed:
Add a custom error and have NewRelic ignore it

We don't want NewRelic to flag deadlock errors in the migration job since we expect them to occur pretty frequently.

anathomical committed:
Fix linting error
